src/unlock/rc_data.rs

```diff
@@ -0,0 +1,205 @@
+use lazy_static::lazy_static;
+
+use sparse_merkle_tree::{default_store::DefaultStore, SparseMerkleTree, H256 as SmtH256};
+
+use crate::types::xudt_rce_mol::{
+    RCDataBuilder, RCDataUnion, RCRuleBuilder, SmtProofBuilder, SmtProofEntryBuilder,
+    SmtProofEntryVec, SmtProofEntryVecBuilder,
+};
+use bytes::Bytes;
+use ckb_hash::{new_blake2b, Blake2b};
+use ckb_types::{molecule, prelude::*};
+use sparse_merkle_tree::traits::Hasher;
+
+lazy_static! {
+    pub static ref SMT_EXISTING: SmtH256 = SmtH256::from([
+        1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0,
+    ]);
+    pub static ref SMT_NOT_EXISTING: SmtH256 = SmtH256::from([
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0,
+    ]);
+}
+
+type SMT = SparseMerkleTree<CKBBlake2bHasher, SmtH256, DefaultStore<SmtH256>>;
+pub type Result<T> = ::core::result::Result<T, RcDataError>;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum RcDataError {
+    BuildTree(String),
+    CompileProof(String),
+}
+
+// on(1): white list
+// off(0): black list
+const WHITE_BLACK_LIST_MASK: u8 = 0x2;
+
+// on(1): emergency halt mode
+// off(0): not int emergency halt mode
+const EMERGENCY_HALT_MODE_MASK: u8 = 0x1;
+pub struct CKBBlake2bHasher(Blake2b);
+
+impl Default for CKBBlake2bHasher {
+    fn default() -> Self {
+        let blake2b = new_blake2b();
+        CKBBlake2bHasher(blake2b)
+    }
+}
+
+impl Hasher for CKBBlake2bHasher {
+    fn write_h256(&mut self, h: &SmtH256) {
+        self.0.update(h.as_slice());
+    }
+    fn finish(self) -> SmtH256 {
+        let mut hash = [0u8; 32];
+        self.0.finalize(&mut hash);
+        hash.into()
+    }
+    fn write_byte(&mut self, b: u8) {
+        self.0.update(&[b][..]);
+    }
+}
+
+fn new_smt(pairs: Vec<(SmtH256, SmtH256)>) -> SMT {
+    let mut smt = SMT::default();
+    for (key, value) in pairs {
+        smt.update(key, value).unwrap();
+    }
+    smt
+}
+
+fn map_exist(hash: &SmtH256) -> (SmtH256, SmtH256) {
+    (*hash, *SMT_EXISTING)
+}
+
+fn map_noexist(hash: &SmtH256) -> (SmtH256, SmtH256) {
+    (*hash, *SMT_NOT_EXISTING)
+}
+
+/// Build the white list smt tree with the given hashes
+/// # Arguments
+/// * `hashes` The given the hashes.
+/// * `on` indicate if the give `hashes` on the list.
+pub fn build_smt_on_wl(hashes: &[SmtH256], on: bool) -> Result<(SmtH256, Vec<u8>)> {
+    let smt_keys = if on { hashes.to_vec() } else { vec![] };
+
+    build_smt_wl(&smt_keys, hashes)
+}
+
+/// Build a white list smt tree with it's keys and gnerate proofs with the according keys.
+/// # Arguments
+/// * `smt_keys` - A list of hashes which to build the smt tree.
+/// * `proof_keys` - The keys to generate the proofs.
+/// # Return
+/// The smt_tree root and the proofs of the proof_keys.
+pub fn build_smt_wl(smt_keys: &[SmtH256], proof_keys: &[SmtH256]) -> Result<(SmtH256, Vec<u8>)> {
```

Comment:
If `proof_keys` is not subset of `smt_keys` it will report `CompileProof`, which will confuse user. We better check `proof_keys` is the subset of `smt_keys` and return a proper error message.